File: RobotWar.spritebuilder/Source/Robot/SimpleRobot.m
//
//  SimpleRobot.m
//  RobotWar
//
//  Created by Benjamin Encz on 03/06/14.
//  Copyright (c) 2014 Apportable. All rights reserved.
//

#import "SimpleRobot.h"

typedef NS_ENUM(NSInteger, RobotAction) {
  RobotActionDefault,
  RobotActionTurnaround
};

@implementation SimpleRobot {
  RobotAction _currentRobotAction;
}

- (void)hitWall:(RobotWallHitDirection)hitDirection hitAngle:(CGFloat)angle {
  [self cancelActiveAction];
  
  _currentRobotAction = RobotActionTurnaround;
  
  switch (hitDirection) {
    case RobotWallHitDirectionFront:
      [self turnRobotRight:180];
      [self moveAhead:20];
      break;
    case RobotWallHitDirectionRear:
      [self moveAhead:80];
      break;
    case RobotWallHitDirectionLeft:
      [self turnRobotRight:90];
      [self moveAhead:20];
      break;
    case RobotWallHitDirectionRight:
      [self turnRobotLeft:90];
      [self moveAhead:20];
      break;
    default:
      break;
  }
  
  _currentRobotAction = RobotActionDefault;
  
}

//- (void)scannedRobot:(Robot *)robot atPosition:(CGPoint)position {
//  if (_currentRobotAction != RobotActionTurnaround) {
//    [self cancelActiveAction];
//    
//    [self turnRobotLeft:20];
//    [self moveBack:80];
//  }
//}

- (void)run {
  while (true) {
    [self moveAhead:80];
    [self turnRobotRight:20];
    [self moveAhead:100];
    [self shoot];
    [self turnRobotLeft:10];
  }
}

- (void)gotHit {
  [self shoot];
  [self turnRobotLeft:45];
  [self moveAhead:100];
}

@end

File: RobotWar.spritebuilder/Source/Robot/SimpleRobot.swift
//
//  SimpleRobotSwift.swift
//  RobotWar
//
//  Created by Dion Larson on 7/2/15.
//  Copyright (c) 2015 MakeGamesWithUs. All rights reserved.
//

import Foundation

class SimpleRobot: Robot {
  
  enum RobotState {                    // enum for keeping track of RobotState
    case Default, Turnaround
  }
  
  var currentRobotState: RobotState = .Default
  
  override func run() {
    while true {
      moveAhead(80)
      turnRobotRight(20)
      moveAhead(100)
      shoot()
      turnRobotLeft(10)
    }
  }
  
  override func scannedRobot(robot: Robot!, atPosition position: CGPoint) {
    // unimplemented
  }
  
  override func gotHit() {
    shoot()
    turnRobotLeft(45)
    moveAhead(100)
  }
  
  override func hitWall(hitDirection: RobotWallHitDirection, hitAngle: CGFloat) {
    cancelActiveAction()
    
    currentRobotState = .Turnaround
    
    switch hitDirection {
    case .Front:
      turnGunRight(180)
      moveAhead(20)
    case .Rear:
      moveAhead(80)
    case .Left:
      turnRobotRight(90)
      moveAhead(20)
    case .Right:
      turnRobotLeft(90)
      moveAhead(20)
    case .None:           // should never be none, but switch must be exhaustive
      break
    }
  }
  
  override func bulletHitEnemy(bullet: Bullet!) {
    // unimplemented
  }
  
}